File: app/utils/engine/CacheEngine.js
import { createClient } from 'redis';

/**
 * Class for performing operations with Redis service
 */
class CacheEngine {
    constructor() {
        const REDIS_HOST = process.env.REDIS_HOST || "localhost";
        const REDIS_PORT = parseInt(process.env.REDIS_PORT) || 6379;

        this.client = createClient({
            socket: {
                host: REDIS_HOST,
                port: REDIS_PORT,
            },
        });

        // Connect to Redis
        this.client.connect()
            .then(() => {
                console.log('Redis client connected to the server');
            })
            .catch(error => {
                console.error(`Redis client error: ${error.message}`);
            });
    }

    /**
     * Checks if connection to Redis is alive
     * @return {boolean} true if connection is alive, false otherwise
     */
    isConnected() {
        return this.client.isOpen;
    }

    /**
     * Gets the value corresponding to the key in Redis
     * @param {string} key - The key to search for in Redis
     * @return {Promise<string>} Value of the key
     */
    async get(key) {
        try {
            return await this.client.get(key);
        } catch (error) {
            console.error(`Error getting key "${key}":`, error);
            throw error;
        }
    }

    /**
     * Creates a new key in Redis with a specific TTL
     * @param {string} key - The key to be saved in Redis
     * @param {string} value - The value to be assigned to the key
     * @param {number} ttl - TTL of the key in seconds
     * @return {Promise<void>} No return
     */
    async set(key, value, ttl = 60 * 24 * 60 * 60) {
        try {
            await this.client.setEx(key, ttl, value);
        } catch (error) {
            console.error(`Error setting key "${key}" with TTL:`, error);
            throw error;
        }
    }

    /**
     * Deletes a key in Redis
     * @param {string} key - The key to be deleted
     * @return {Promise<void>} No return
     */
    async del(key) {
        try {
            await this.client.del(key);
        } catch (error) {
            console.error(`Error deleting key "${key}":`, error);
            throw error;
        }
    }

    /**
     * Increases a key's value in Redis
     * @param {string} key - The key to increment
     * @return {Promise<number>} Updated value of the key
     */
    async incr(key) {
        try {
            return await this.client.incr(key);
        } catch (error) {
            console.error(`Error incrementing key "${key}":`, error);
            throw error;
        }
    }

    /**
     * Make a key expire in Redis
     * @param {string} key - The key to be expired
     * @param {number} ttl - The time window to expire in seconds
     * @return {Promise<void>} No return
     */
    async expire(key, ttl) {
        try {
            await this.client.expire(key, ttl);
        } catch (error) {
            console.error(`Error setting expiration for key "${key}":`, error);
            throw error;
        }
    }

    /**
     * Measures the latency to the Redis server.
     * @returns {Promise<number>} - The latency in milliseconds.
     */
    async getLatency() {
        try {
            const start = Date.now();
            await this.client.ping();
            return Date.now() - start;
        } catch (error) {
            console.error("Failed to measure latency:", error);
            throw error;
        }
    }

    /**
     * Saves an object to Redis
     * @param {string} key - The key to be saved in Redis
     * @param {Object} value - The object to be assigned to the key
     * @param {number} ttl - TTL of the key in seconds
     * @return {Promise<void>} No return
     */
    async setObject(key, value, ttl = 60 * 24 * 60 * 60) {
        try {
            const stringValue = JSON.stringify(value);
            await this.client.setEx(key, ttl, stringValue);
        } catch (error) {
            console.error(`Error setting object for key "${key}" with TTL:`, error);
            throw error;
        }
    }

    /**
     * Gets an object from Redis
     * @param {string} key - The key to search for in Redis
     * @return {Promise<Object>} The object corresponding to the key
     */
    async getObject(key) {
        try {
            const value = await this.client.get(key);
            return JSON.parse(value);
        } catch (error) {
            console.error(`Error getting object for key "${key}":`, error);
            throw error;
        }
    }
}

const cacheEngine = new CacheEngine();

export {cacheEngine};
export default CacheEngine;

File: app/utils/engine/CacheEngine.py
import os
import redis
import json

class CacheEngine:
    def __init__(self):
        REDIS_HOST = os.getenv('REDIS_HOST', 'localhost')
        REDIS_PORT = int(os.getenv('REDIS_PORT', 6379))

        self.client = redis.Redis(host=REDIS_HOST, port=REDIS_PORT)

        try:
            self.client.ping()
            print('Redis client connected to the server')
        except redis.ConnectionError as error:
            print(f'Redis client error: {error}')

    def is_connected(self):
        return self.client.ping()

    async def get(self, key):
        try:
            return self.client.get(key)
        except redis.RedisError as error:
            print(f'Error getting key "{key}": {error}')
            raise error

    async def set(self, key, value, ttl=60 * 24 * 60 * 60):
        try:
            self.client.setex(key, ttl, value)
        except redis.RedisError as error:
            print(f'Error setting key "{key}" with TTL: {error}')
            raise error

    async def del_key(self, key):
        try:
            self.client.delete(key)
        except redis.RedisError as error:
            print(f'Error deleting key "{key}": {error}')
            raise error

    async def incr(self, key):
        try:
            return self.client.incr(key)
        except redis.RedisError as error:
            print(f'Error incrementing key "{key}": {error}')
            raise error

    async def expire(self, key, ttl):
        try:
            self.client.expire(key, ttl)
        except redis.RedisError as error:
            print(f'Error setting expiration for key "{key}": {error}')
            raise error

    async def get_latency(self):
        try:
            start = time.time()
            self.client.ping()
            return (time.time() - start) * 1000
        except redis.RedisError as error:
            print(f'Failed to measure latency: {error}')
            raise error

    async def set_object(self, key, value, ttl=60 * 24 * 60 * 60):
        try:
            string_value = json.dumps(value)
            self.client.setex(key, ttl, string_value)
        except redis.RedisError as error:
            print(f'Error setting object for key "{key}" with TTL: {error}')
            raise error

    async def get_object(self, key):
        try:
            value = self.client.get(key)
            return json.loads(value)
        except redis.RedisError as error:
            print(f'Error getting object for key "{key}": {error}')
            raise error

cache_engine = CacheEngine()
